2016/presentations/Debugging_of_CPython_processes_with_gdb/text.md

```diff
@@ -0,0 +1,639 @@
+#10 PyEval_EvalFrameEx () at ../Python/ceval.c:2987
+#11 0x00000000004c32e5 in PyEval_EvalCodeEx () at ../Python/ceval.c:3582
+#12 0x00000000004c3089 in PyEval_EvalCode (co=<optimized out>, globals=<optimized out>, locals=<optimized out>) at ../Python/ceval.c:669
+#13 0x00000000004f263f in run_mod.lto_priv () at ../Python/pythonrun.c:1376
+#14 0x00000000004ecf52 in PyRun_FileExFlags () at ../Python/pythonrun.c:1362
+#15 0x00000000004eb6d1 in PyRun_SimpleFileExFlags () at ../Python/pythonrun.c:948
+#16 0x000000000049e2d8 in Py_Main () at ../Modules/main.c:640
+#17 0x00007f3e94fc2610 in __libc_start_main (main=0x49dc00 <main>, argc=2, argv=0x7ffefb504c98, init=<optimized out>, fini=<optimized out>, 
+    rtld_fini=<optimized out>, stack_end=0x7ffefb504c88) at libc-start.c:291
+#18 0x000000000049db29 in _start ()
+
+(gdb) py-bt
+
+Undefined command: "py-bt".  Try "help".
+
+```
+
+Once again, this is caused by the fact that `python` binary in a virtual
+environment has a different path. By default, `gdb` will try to [auto-load]
+Python extensions for a particular object file under debug, if they exist.
+Specifically, `gdb` will look for `objfile-gdb.py` and try to `source` it on
+start:
+
+```
+
+(gdb) info auto-load
+
+gdb-scripts:  No auto-load scripts.
+libthread-db:  No auto-loaded libthread-db.
+local-gdbinit:  Local .gdbinit file was not found.
+python-scripts:
+Loaded  Script
+Yes     /usr/share/gdb/auto-load/usr/bin/python2.7-gdb.py
+
+```
+
+If, for some reason this has not been done, you can always do it manually:
+
+```
+
+(gdb) source /usr/share/gdb/auto-load/usr/bin/python2.7-gdb.py
+
+```
+
+e.g. if you want to test a new version of the `gdb` extension shipped with CPython.
+
+[separate]: https://sourceware.org/gdb/onlinedocs/gdb/Separate-Debug-Files.html
+[ELF]: https://en.wikipedia.org/wiki/Executable_and_Linkable_Format
+[auto-load]: https://sourceware.org/gdb/onlinedocs/gdb/Python-Auto_002dloading.html#set%20auto%2dload%20python%2dscripts
+
+
+## PyPy, Jython, etc
+
+The described debugging technique is only feasible for the CPython interpreter
+as is, as the `gdb` extension is specifically written to introspect the state
+of CPython internals (e.g. `PyEval_EvalFrameEx` calls).
+
+For [PyPy] there is an open [issue] on Bitbucket, where it was proposed to
+provide integration with `gdb`, but looks like the attached patches have not
+been merged yet and the person, who wrote those, lost interest in this.
+
+For [Jython] you could probably use standard tools for debugging of `JVM`
+applications, e.g. [VisualVM].
+
+[issue]: https://bitbucket.org/pypy/pypy/issues/1204/gdb-hooks-for-debugging-pypy
+[PyPy]: http://pypy.org/
+[Jython]: http://www.jython.org/
+[VisualVM]: http://visualvm.java.net/
+
+
+Conclusion
+----------
+
+`gdb` is a powerful tool, that allows one to debug complex problems with
+crashing or hanging CPython processes, as well as Python code, that does
```

Comment:
`gdb` is a powerful tool that allows one to debug complex problems with
 crashing or hanging CPython processes, as well as Python code that does  (spurious commas)